pyproject.toml

```diff
@@ -0,0 +1,100 @@
+[build-system]
+requires = ["poetry>=0.12"]
+build-backend = "poetry.masonry.api"
+
+[tool.poetry]
+name = "relate-courseware"
+version = "2016.1"
+description = "RELATE courseware"
+readme = "README.rst"
+repository = "https://github.com/inducer/relate"
+documentation = "https://documen.tician.de/relate/"
+authors = ["Andreas Kloeckner <inform@tiker.net>"]
+license = "MIT"
+packages = [
+    { include = "course" },
+    { include = "accounts" },
+    { include = "relate" },
+    { include = "bin" },
+]
+
+[tool.poetry.dependencies]
+python = "^3.7"
+django = "^3.0.7"
+# Automatically renders Django forms in a pretty, Bootstrap-compatible way.
+django-crispy-forms = ">=1.5.1"
+# Page data, answer data, ... all represented in JSON. This makes that editable in the Django admin.
+jsonfield = ">=1.4.0"
+# /!\ Upstream is dead, using branch for Django 3.0 support
+django-yamlfield = {git = "https://github.com/bakatrouble/django-yamlfield.git", rev = "c92d0373d12a02d1e52fb09b44010f156111d7ea"}
+# For easy content formatting:
+markdown = "^2.6.3"
+# For rendering macros in content:
+jinja2 = "^2.11.2"
+# For math/symbolic questions
+pymbolic = "*"
+sympy = "*"
+# Django timezone support
+pytz = "^2020.1"
+# Course content is stored in YAML.
+pyyaml = "*"
+# Dulwich git integration
+dulwich = "^0.19"
+ecdsa = "^0.15"
+# FIXME: 2.7.1 gives bogus 'incorrect padding' error on PEM parsing
+paramiko = "<2.7"
+# A date picker widget - https://github.com/tutorcruncher/django-bootstrap3-datetimepicker
+django-bootstrap3-datetimepicker-2 = "^2.8.2"
+# For in-class instant messaging
+dnspython = "^1.16.0"
+SleekXMPP = "^1.3.3"
+# To manage web dependencies
+django-npm = "^1.0.0"
+# For comfortable code entry
+django-codemirror-widget = ">=0.5"
+# For code isolation in code questions
+docker-py = "^1.10.6"
+# For code highlighting, required via the CodeHilite extension to Python-Markdown
+pygments = "^2.6.1"
+# For grade export
+unicodecsv = "^0.14.1"
+# To support network matching for facility recognition
+ipaddress = "^1.0.23"
```

Comment:
My mental model was that, in addition to these version ranges, there would be something like a "lock file" (like with npm, also in pipenv), that specifies exact version numbers of all dependencies that were tested together (for more predictable deploys). Does poetry allow such a thing?

Reply:
There is, I just haven't committed it yet since there is a lot of churn. I will as soon as its not changing drastically and things are ready to merge